docs/packages/cluster/AUTOSCALING.md

```diff
@@ -0,0 +1,138 @@
+| 30s              | 4             | 2000MB                | 1500MB              |
+
+We end up with two extra servers to provision atop our original 2.
+
+```rust
+ceil((3000 - 1500) / 1000) = 2
+```
+
+| time since start | desired count | expected total memory | actual total memory |
+| ---------------- | ------------- | --------------------- | ------------------- |
+| 0s               | 2             | 2000MB                | 0MB                 |
+| 15s              | 3             | 2000MB                | 1000MB              |
+| 30s              | 4             | 2000MB                | 1500MB              |
+| 45s              | 4             | 2000MB                | 2000MB              |
+
+And finally we reach the desired capacity.
+
+### Failover has higher specs
+
+In the event that the failover hardware has higher specs than the desired amount, there is no error system that reduces the desired count to account for this difference. This is because there is no direct correlation between desired count and the hardware being provisioned and destroyed. Thus, if hardware with higher than expected specs is provisioned, that extra space will not be taken into account.
+
+If it was taken into account in a similar error system as failover with lower specs, it would look like this:
+
+| time since start | desired count | expected total memory | actual total memory |
+| ---------------- | ------------- | --------------------- | ------------------- |
+| 0s               | 1             | 1000MB                | 2000MB              |
+
+Error:
+
+```rust
+ceil(expected - actual) / expected_memory_per_server)
+
+ceil((1000 - 2000) / 1000) = -1
+```
+
+The original desired count + error would be 0, destroying the only server and causing the capacity to drop to 0. If the higher-spec'd failover kept getting provisioned, this would end up in a loop.
+
+## Job server autoscaling
+
+The nomad topology for each job server in a datacenter is fetched and the memory is aggregated. This value is then divided by the expected memory capacity (the capacity of the first hardware in the config), which determines the minimum expected server count required to accommodate the current usage. Then, we add the error value (discussed above) and the margin value which is configured in the namespace config.
+
+### Autoscaling via machine learning
+
+Coming soon
+
+## GG server autoscaling
+
+Because we do not need to be preemptive with GG servers, the autoscaling is a bit more simple.
+
+-   If the current CPU usage is more than 20% under the total, add a server.
+-   If the current CPU usage is less than 130% under the total, remove a server.
+
+Examples:
+
+```rust
+// 3 servers
+total_cpu = 300
+cpu_usage = 285
+
+// result: add a server
+```
+
+```rust
+// 1 server
+total_cpu = 100
+cpu_usage = 70
+
+// result: do nothing
+```
+
+```rust
+// 4 servers
+total_cpu = 400
+cpu_usage = 250
+
+// result: remove a server
+```
```

Comment:
Might be nice to have the percents:

````suggestion
```rust
// 3 servers
total_cpu = 300
cpu_usage = 285

// result: 95% of the total, add a server
```

```rust
// 1 server
total_cpu = 100
cpu_usage = 70

// result: 70% of the total, do nothing
```

```rust
// 4 servers
total_cpu = 400
cpu_usage = 250

// result: 63% of the total, remove a server
```
````

Reply:
See my above comment, scaling down isn't based on percent of total.